Update celery to latest version

AmitPhulera opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
This is an attempt to update celery to the latest version. There are a lot of changes/deprecations that were done in celery >= 5.0.0. The changes that affect us are - 
- `celery.task.task` was replace with `celery.shared_task` (`celery.task` was completely removed).
- `periodic_task` was completely removed. So we have to write our own version of it. See 7e7701eb832b587c3b2e843d3e49c5984fe05ef2
- `celery.task.revoke` was moved to `celery.app.control.Control.revoke`
- `kombu.five` was removed in the latest version of kombu as an effort to drop support for python3. We have been using it for a hack https://github.com/dimagi/commcare-hq/blob/master/corehq/util/celery_utils.py#L42 which we don't need now.

Review by commit 🐟 

## Technical Summary 
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
[JIRA ticket](https://dimagi-dev.atlassian.net/browse/SAAS-13487)
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance
The PR will be tested on staging and would go through QA.
### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

AmitPhulera committed:
make requirements: update all requirement files

AmitPhulera committed:
add: periodic_task decorator and PeriodicTask class

AmitPhulera committed:
replace: celery.task.periodic_task with corehq.apps.celery.periodic_task

AmitPhulera committed:
replace: celery.task.task with celery.shared_task

AmitPhulera committed:
remove: unused imports

AmitPhulera committed:
remove: @task with @shared_task at other places

AmitPhulera committed:
update: revoking task logic

AmitPhulera committed:
remove: kombu.five

time_start is no longer montonic https://stackoverflow.com/a/20096342/3537212, so we can use it directly without any hacks

AmitPhulera committed:
update: handle cases where no option in passed

AmitPhulera commented on corehq/util/celery_utils.py:
```diff
@@ -39,9 +39,8 @@ class TaskInfo(object):
-        # http://stackoverflow.com/questions/20091505/celery-task-with-a-time-start-attribute-in-1970
         if time_start:
-            self.time_start = datetime.fromtimestamp(time() - kombu.five.monotonic() + time_start)
+            self.time_start = datetime.fromtimestamp(time_start)
```
@calellowitz  @millerdev Just want to be sure that my understanding of it is correct.
`time_start` now no longer is montonic in celery. See the comment http://stackoverflow.com/questions/20091505/celery-task-with-a-time-start-attribute-in-1970

millerdev replied:
I don't think I have enough context to answer that or whether it matters. We should probably test this to ensure it does the right thing with the new version of Celery.

millerdev commented on requirements/dev-requirements.txt:
```diff
@@ -145,7 +155,7 @@ django-formtools==2.3
     #   django-two-factor-auth
 django-logentry-admin==1.0.6
     # via -r base-requirements.in
-django-nose @ https://github.com/dimagi/django-nose/raw/fast-first-1.4.6.1/releases/django_nose-1.4.6.1-py2.py3-none-any.whl
+https://github.com/dimagi/django-nose/raw/fast-first-1.4.6.1/releases/django_nose-1.4.6.1-py2.py3-none-any.whl
```
I don't think this should have changed. What version of pip-tools did you use to make requirements? I have version 6.6.0

AmitPhulera replied:
Yep. I had an older version of pip-tools. Fixed it https://github.com/dimagi/commcare-hq/pull/31584/commits/9653a10c4258708538534ccd25de8c06c8d043cf

millerdev commented on corehq/apps/celery/__init__.py:
```diff
@@ -18,3 +18,25 @@ def _init_celery_app():
+        }
+
+
+def periodic_task(*args, **options):
```
Why `*args`? It doesn't appear to be used.

AmitPhulera replied:
Yeah. Only `**options` are used. I kept `*args` to make it consistent with the implementation that celery used to have https://github.com/celery/celery/blob/b2668607c909c61becd151905b4525190c19ff4a/celery/task/base.py#L275 

Since we are not using it, I think it would be a good idea to remove it.

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/31584/commits/0403bc088115d41c7eb35e3a0143808c48a1218c

millerdev commented on corehq/apps/celery/__init__.py:
```diff
@@ -18,3 +18,25 @@ def _init_celery_app():
+
+    @classmethod
+    def on_bound(cls, app):
+        app.conf.beat_schedule[cls.name] = {
```
Will this blow away a previous schedule if there was one for `cls.name`? Should it verify that there was no previous schedule, maybe add `assert cls.name not in app.conf.beat_schedule, cls.name` above this line?

AmitPhulera replied:
I tested it locally. It does not blow up. It just ignores any other task with the same name.
If there are tasks with the same names in different task files, I guess their `cls.name` would be different. So assert would not catch it. And if we define two tasks with the same name in the same task file then the stickler will complain about function redefinition. 
It would not harm to add assert but wondering if it would it be beneficial?

millerdev commented on corehq/apps/celery/__init__.py:
```diff
@@ -18,3 +18,25 @@ def _init_celery_app():
+            'queue': options.pop('queue')
+        }
+    options['base'] = PeriodicTask
+    return app.task(**options)
```
I think these should live in a different module. Suggestion: `corehq.apps.celery.periodic`. And `periodic_task` should be imported in this module (`__init__.py`) to make it easier to find in other places.

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/31584/commits/28ea3e2ebfdb4c80801778ef53f95b930e8be9b8  https://github.com/dimagi/commcare-hq/pull/31584/commits/d0187705527d0e5c123facedfef63472c52e290e

millerdev commented:
The label could probably be _product/invisible_ rather than _product/all-users-all-environments_. I assume we will do extensive QA testing on this upgrade, is that right? (Currently there is nothing specified for QA Plan.)

AmitPhulera committed:
fix: run make requirements with latest pip-tools

AmitPhulera committed:
update: warnings.py

AmitPhulera committed:
move: periodic_task functionality to corehq.apps.celery.periodic

AmitPhulera committed:
remove: periodic_task functionality from __init__.py

AmitPhulera committed:
fix: tests

AmitPhulera committed:
remove: *args from periodict task

AmitPhulera committed:
add: docstring for periodic tasks

AmitPhulera committed:
add: corehq.apps.userreports.util to warninigs.py

AmitPhulera commented:
@millerdev I have updated the flag. 
And yes we are planning to do an extensive QA on it. I was waiting on getting this PR QAable so that no major changes are required in it.

AmitPhulera committed:
update: django-celery-result migrations in lock file

AmitPhulera committed:
Apply suggestions from code review

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

AmitPhulera commented on corehq/apps/celery/periodic.py:
```diff
@@ -0,0 +1,44 @@
+        })
+    options['base'] = PeriodicTask
+    from corehq.apps.celery import app
+    return app.task(**options)
```
Just curious - Why does the suggestion seem more apt to you?

AmitPhulera commented on corehq/warnings.py:
```diff
@@ -32,17 +31,18 @@
     ("", "json_response is deprecated.  Use django.http.JsonResponse instead."),
     ("", "property_match are deprecated. Use boolean_expression instead."),
     ("corehq.util.validation", "metaschema specified by $schema was not found"),
+    ("corehq.apps.userreports.util", "'collections.abc'"),
```
It s used here - https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/userreports/util.py#L35

I will sort them.

millerdev replied:
I don't think that usage would cause that error. `collections.Mapping` is not `collections.abc`, and should be fine to reference.

AmitPhulera commented on corehq/apps/accounting/tasks.py:
```diff
@@ -15,7 +15,7 @@
 import six.moves.urllib.request
 from celery.schedules import crontab
-from celery.task import periodic_task, task
+from celery import shared_task
```
Thanks for the suggestion. I also had something simpler in my mind where we just reference `shared_task` as `task` in the imports.

```
from celery import shared_task as task
```
But I wanted to keep the implementation closer to celery API. I can update the PR to use the patched code but again this would be an additional knowledge that we would have to maintain.

AmitPhulera replied:
I missed your comment earlier @dannyroberts, apologies for that.
I understand your concern, however, the frequency with which task files are updated would be very less. But I would raise another PR with minimal changes and rebase this PR with the master to ensure there are no conflicts.

AmitPhulera committed:
Merge branch 'ap/celery-latest-version' of https://github.com/dimagi/commcare-hq into ap/celery-latest-version

AmitPhulera committed:
add: CELERY_RESULT_SERIALIZER to pickle to test for test failures

AmitPhulera committed:
remove: celery_result_serializer as pickle

AmitPhulera committed:
fix: tests by removing dependecny of timingcontext

In the newer versions of celery the args that are passed in and rfirst serialized and then retrieved which changes the objects so timing context does not work

AmitPhulera committed:
merge: conflicts fix

millerdev commented on corehq/ex-submodules/casexml/apps/phone/tests/test_async_restore.py:
```diff
@@ -146,16 +147,17 @@ def test_completed_task_deletes_cache(self):
-        self.assertTrue(restore_config.timing_context.is_finished())
-        self.assertIsNotNone(restore_config.restore_state.current_sync_log)
+        sync_logs = SyncLogSQL.objects.filter(domain='dummy-project')
+        self.assertEqual(len(sync_logs), 1)
```
Is it true that the number of sync logs before this test is run should be zero? What do you think of adding a check for that as well?

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/31584/commits/b1e66c1a4e15b4603dc22bcefb1c5948d2cbb833

AmitPhulera committed:
update tests:

AmitPhulera commented:
Closing this as it is superseded by https://github.com/dimagi/commcare-hq/pull/32134
